File: integration/test/test_profile_policy.py
#!/usr/bin/env python3
#
#  Copyright (c) 2015 - 2023, Intel Corporation
#  SPDX-License-Identifier: BSD-3-Clause
#


import os
import sys
import pandas
import signal
import unittest
import subprocess

import geopmpy.launcher
import geopmpy.io
import geopmpy.policy_store

from integration.test import util


@util.skip_unless_do_launch()
@util.skip_unless_batch()
class TestIntegrationProfilePolicy(unittest.TestCase):
    def setUp(self):
        # clean up stale keys
        try:
            os.unlink("/dev/shm/geopm*")
        except:
            pass

        self._files = []
        self.default_power_cap = 142
        self.custom_power_cap = 152
        # file name must match name in .cpp file
        script_dir = os.path.dirname(os.path.realpath(__file__))
        exe_path = os.path.join(script_dir, '.libs', 'test_profile_policy')
        policy_db_path = os.path.join(os.getcwd(), 'policystore.db')
        self._files.append(policy_db_path)
        geopmpy.policy_store.connect(policy_db_path)
        geopmpy.policy_store.set_default("power_balancer",
                                         [self.default_power_cap])
        geopmpy.policy_store.set_best("power_balancer", "power_custom",
                                      [self.custom_power_cap, 0, 0, 0])
        geopmpy.policy_store.disconnect()

        # common run parameters
        self._num_rank = 1
        self._num_node = 1
        agent = 'power_balancer'
        self._app_conf = geopmpy.io.BenchConf('test_profile_policy_app.config')
        self._app_conf.append_region('sleep', 1.0)
        self._app_conf.write()
        self._files.append(self._app_conf.get_path())
        # must match prefix in .cpp file
        endpoint_prefix = '/geopm_endpoint_profile_policy_test'
        # test launcher sets profile, have to use real launcher for now
        self._argv = ['dummy', 'srun',
                      '--geopm-endpoint', endpoint_prefix,
                      '--geopm-agent', agent,
                      '--geopm-timeout', "5"]

        # this must be launched on the same node as the root controller
        # for now limit test to one node
        self._endpoint_mgr = subprocess.Popen([exe_path],
                                              stdout=subprocess.PIPE,
                                              stderr=subprocess.PIPE)

    def tearDown(self):
        # kill policy handler in case it didn't exit on its own
        self._endpoint_mgr.kill()
        sys.stdout.write(self._endpoint_mgr.stdout.read().decode())
        sys.stdout.write(self._endpoint_mgr.stderr.read().decode())

        # clean up shared memory
        try:
            os.unlink("/dev/shm/geopm*")
        except:
            pass

    @util.skip_unless_config_enable('beta')
    def test_policy_default(self):
        profile = 'unknown'
        report_path = profile + '.report'
        policy_trace = profile + '.trace-policy'
        self._files.append(report_path)
        self._files.append(policy_trace)
        self._argv.extend(['--geopm-profile', profile])
        self._argv.extend(['--geopm-trace-endpoint-policy', policy_trace])
        self._argv.extend(['--geopm-report', report_path])
        self._argv.append(self._app_conf.get_exec_path())
        self._argv.extend(self._app_conf.get_exec_args())
        launcher = geopmpy.launcher.Factory().create(self._argv, self._num_rank, self._num_node)
        launcher.run()

        report_data = geopmpy.io.RawReport(report_path).meta_data()
        self.assertEqual(report_data['Profile'], profile)
        policy = report_data['Policy']
        self.assertEqual(policy, 'DYNAMIC')
        # check profile trace for single line with this power cap
        csv_data = pandas.read_csv(policy_trace, delimiter='|', comment='#')
        self.assertEqual(csv_data['CPU_POWER_LIMIT'][0], self.default_power_cap)

    @util.skip_unless_config_enable('beta')
    def test_policy_custom(self):
        profile = 'power_custom'
        report_path = profile + '.report'
        policy_trace = profile + '.trace-policy'
        self._files.append(report_path)
        self._files.append(policy_trace)
        self._argv.extend(['--geopm-profile', profile])
        self._argv.extend(['--geopm-trace-endpoint-policy', policy_trace])
        self._argv.extend(['--geopm-report', report_path])
        self._argv.append(self._app_conf.get_exec_path())
        self._argv.extend(self._app_conf.get_exec_args())
        launcher = geopmpy.launcher.Factory().create(self._argv, self._num_rank, self._num_node)
        launcher.run()

        report_data = geopmpy.io.RawReport(report_path).meta_data()
        self.assertEqual(report_data['Profile'], profile)
        policy = report_data['Policy']
        self.assertEqual(policy, 'DYNAMIC')
        # check profile trace for single line with this power cap
        csv_data = pandas.read_csv(policy_trace, delimiter='|', comment='#')
        self.assertEqual(csv_data['CPU_POWER_LIMIT'][0], self.custom_power_cap)


if __name__ == '__main__':
    unittest.main()

File: integration/test/test_profile_policy.cpp
/*
 * Copyright (c) 2015 - 2024, Intel Corporation
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include <signal.h>

#include <string>

#include "geopm_daemon.h"
#include "geopm/Exception.hpp"


struct geopm_daemon_c *g_daemon;
static void handler(int sig)
{
    if (g_daemon) {
        geopm_daemon_stop_wait_loop(g_daemon);
    }
}

int main(int argc, char* argv[])
{
    // Handle signals
    struct sigaction act;
    act.sa_handler = handler;
    sigaction(SIGINT, &act, NULL);

    std::string db_path = "policystore.db";
    std::string endpoint_name = "/geopm_endpoint_profile_policy_test";

    int err = 0;
    err = geopm_daemon_create(endpoint_name.c_str(),
                              db_path.c_str(),
                              &g_daemon);
    if (err) {
        throw geopm::Exception("geopm_daemon_create() failed",
                               GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
    }
    err = geopm_daemon_update_endpoint_from_policystore(g_daemon, 10);
    if (err) {
        geopm_daemon_destroy(g_daemon);
        throw geopm::Exception("geopm_daemon_update_endpoint_from_policystore() failed",
                               GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
    }
    err = geopm_daemon_destroy(g_daemon);
    if (err) {
        throw geopm::Exception("geopm_daemon_daemon() failed",
                               GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
    }

    // cpp version: include "geopm/Daemon.hpp"
    //auto daemon = geopm::Daemon::make_unique(endpoint_name, db_path);
    //daemon->update_endpoint_from_policystore(10);

    return 0;
}
